File: app/src/main/java/dev/bee/kanjianki/MainActivitySettingsAnkiSourceImportFilters.java
package dev.bee.kanjianki;

import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Toast;

import dev.bee.kanjianki.core.RecordsBase;
import dev.bee.kanjianki.core.RecordsSyncModels;
import dev.bee.kanjianki.core.SettingsImportPreset;
import dev.bee.kanjianki.core.SettingsTextCopy;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

final class MainActivitySettingsAnkiSourceImportFilters {
    private final MainActivitySettings activity;
    private final MainActivitySettingsAnkiSourceInputs inputs;
    private final MainActivitySettingsAnkiSourceValidation validation;

    MainActivitySettingsAnkiSourceImportFilters(
            MainActivitySettings activity,
            MainActivitySettingsAnkiSourceInputs inputs,
            MainActivitySettingsAnkiSourceValidation validation
    ) {
        this.activity = activity;
        this.inputs = inputs;
        this.validation = validation;
    }

    View importFilterSettingsPanel(RecordsSyncModels.Settings current) {
        CheckBox activeCards = activity.importFilterCheckBox(SettingsTextCopy.activeCardsLabel(), current.importActiveCards);
        CheckBox suspendedCards = activity.importFilterCheckBox(SettingsTextCopy.suspendedCardsLabel(), current.importSuspendedCards);
        CheckBox taggedCards = activity.importFilterCheckBox(SettingsTextCopy.taggedCardsLabel(), current.importTaggedCardsEnabled());
        CheckBox weakCards = activity.importFilterCheckBox(SettingsTextCopy.weakCardsLabel(), current.importWeakCards);
        CheckBox browserQueryCards = activity.importFilterCheckBox(SettingsTextCopy.browserQueryLabel(), current.importBrowserQueryCards);

        EditText browserQueryInput = inputs.fieldInput(current.importBrowserQuery);
        browserQueryInput.setHint(SettingsTextCopy.ankiBrowserQueryHint());
        browserQueryInput.setContentDescription(SettingsTextCopy.ankiBrowserQueryLabel());

        EditText tags = inputs.fieldInput(current.importTagsText());
        tags.setHint(SettingsTextCopy.ankiNoteTagsHint());
        tags.setContentDescription(SettingsTextCopy.ankiNoteTagsLabel());

        EditText difficultyInput = inputs.decimalInput(current.importWeakFsrsDifficultyThreshold);
        difficultyInput.setContentDescription(SettingsTextCopy.fsrsDifficultyLabel());
        EditText lapses = activity.thresholdInput(current.importWeakLapsesThreshold);
        lapses.setContentDescription(SettingsTextCopy.lapsesLabel());

        EditText minMatching = activity.thresholdInput(current.importMinMatchingCardsPerKanji);
        minMatching.setContentDescription(SettingsTextCopy.minimumMatchingCardsLabel());

        return MainActivitySettingsAnkiSourceImportFiltersCompose.importFiltersSettingsPanelView(
                activity,
                new SettingsImportFiltersPanelModel(
                        SettingsTextCopy.importFiltersTitle(),
                        SettingsTextCopy.settingsImportSummary(current),
                        SettingsTextCopy.importFiltersBody(),
                        SettingsTextCopy.presetsTitle(),
                        presetButtons(),
                        Arrays.asList(activeCards, suspendedCards, taggedCards, weakCards, browserQueryCards),
                        fieldModel(SettingsTextCopy.ankiBrowserQueryLabel(), browserQueryInput, 52),
                        fieldModel(SettingsTextCopy.ankiNoteTagsLabel(), tags, 52),
                        fieldModel(SettingsTextCopy.fsrsDifficultyLabel(), difficultyInput, 58),
                        fieldModel(SettingsTextCopy.lapsesLabel(), lapses, 58),
                        fieldModel(SettingsTextCopy.minimumMatchingCardsLabel(), minMatching, 58),
                        SettingsTextCopy.saveImportFiltersLabel(),
                        () -> saveImportFilters(
                                activeCards,
                                suspendedCards,
                                taggedCards,
                                weakCards,
                                browserQueryCards,
                                tags,
                                browserQueryInput,
                                difficultyInput,
                                lapses,
                                minMatching
                        )
                )
        );
    }

    private List<SettingsImportPresetButtonModel> presetButtons() {
        List<SettingsImportPresetButtonModel> models = new ArrayList<>();
        for (SettingsImportPreset preset : SettingsImportPreset.defaults()) {
            models.add(new SettingsImportPresetButtonModel(preset.label(), () -> applyPreset(preset)));
        }
        return models;
    }

    private SettingsImportFilterFieldModel fieldModel(String label, EditText input, int heightDp) {
        return new SettingsImportFilterFieldModel(label, input, heightDp);
    }

    private void saveImportFilters(
            CheckBox activeCards,
            CheckBox suspendedCards,
            CheckBox taggedCards,
            CheckBox weakCards,
            CheckBox browserQueryCards,
            EditText tags,
            EditText browserQueryInput,
            EditText difficultyInput,
            EditText lapses,
            EditText minMatching
    ) {
        List<String> parsedTags = RecordsBase.parseImportTags(tags.getText().toString());
        String queryText = browserQueryInput.getText().toString().trim();
        if (browserQueryCards.isChecked() && queryText.isEmpty()) {
            Toast.makeText(activity, SettingsTextCopy.browserQueryRequiredToast(), Toast.LENGTH_SHORT).show();
            return;
        }
        if (!validation.hasSelectedImportSource(activeCards, suspendedCards, taggedCards, weakCards, browserQueryCards, parsedTags, queryText)) {
            Toast.makeText(activity, SettingsTextCopy.importSourceRequiredToast(), Toast.LENGTH_SHORT).show();
            return;
        }
        MainActivityBase.ImportThresholds parsedThresholds = validation.readImportThresholds(difficultyInput, lapses, minMatching);
        if (parsedThresholds == null) {
            return;
        }
        SettingsWriteActions.saveImportFilters(
                new SettingsWriteActions.ImportFilterWriteRequest(
                        activeCards.isChecked(),
                        suspendedCards.isChecked(),
                        taggedCards.isChecked(),
                        String.join(" ", parsedTags),
                        weakCards.isChecked(),
                        parsedThresholds.difficulty,
                        parsedThresholds.lapseThreshold,
                        parsedThresholds.minCards,
                        browserQueryCards.isChecked(),
                        queryText
                ),
                new MainActivitySettingsAnkiSourceWriter(activity)
        );
        Toast.makeText(activity, SettingsTextCopy.importFiltersSavedToast(), Toast.LENGTH_LONG).show();
        activity.renderSettings();
    }

    private void applyPreset(SettingsImportPreset preset) {
        SettingsWriteActions.saveImportFilters(
                new SettingsWriteActions.ImportFilterWriteRequest(
                        preset.activeCards(),
                        preset.suspendedCards(),
                        preset.taggedCards(),
                        preset.tags(),
                        preset.weakCards(),
                        preset.weakDifficulty(),
                        preset.weakLapses(),
                        preset.minMatchingCards(),
                        preset.browserQueryCards(),
                        preset.browserQuery()
                ),
                new MainActivitySettingsAnkiSourceWriter(activity)
        );
        Toast.makeText(activity, SettingsTextCopy.importPresetSavedToast(), Toast.LENGTH_LONG).show();
        activity.renderSettings();
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/MainActivitySettingsAnkiSourceImportFilters.kt
package dev.bee.kanjianki

import android.widget.Toast
import dev.bee.kanjianki.core.RecordsBase
import dev.bee.kanjianki.core.RecordsSyncModels
import dev.bee.kanjianki.core.SettingsImportPreset
import dev.bee.kanjianki.core.SettingsTextCopy
import dev.bee.kanjianki.data.SettingsSaveCommand
import java.util.Locale

internal class MainActivitySettingsAnkiSourceImportFilters(
    private val activity: MainActivitySettings,
    private val validation: MainActivitySettingsAnkiSourceValidation,
) {
    fun importFilterSettingsPanelModel(
        current: RecordsSyncModels.Settings,
        tagRepairedCards: Boolean,
    ): SettingsImportFiltersPanelModel {
        val state = SettingsImportFiltersState(
            activeCards = current.importActiveCards,
            suspendedCards = current.importSuspendedCards,
            taggedCards = current.importTaggedCardsEnabled(),
            weakCards = current.importWeakCards,
            browserQueryCards = current.importBrowserQueryCards,
            browserQuery = trimmed(current.importBrowserQuery),
            tags = trimmed(current.importTagsText()),
            difficulty = decimalText(current.importWeakFsrsDifficultyThreshold),
            lapses = thresholdText(current.importWeakLapsesThreshold),
            minMatching = thresholdText(current.importMinMatchingCardsPerKanji),
            tagRepairedCards = tagRepairedCards,
        )

        return SettingsImportFiltersPanelModel(
            title = SettingsTextCopy.importFiltersTitle(),
            summary = SettingsTextCopy.settingsImportSummary(current),
            body = SettingsTextCopy.importFiltersBody(),
            presetsTitle = SettingsTextCopy.presetsTitle(),
            presets = presetButtons(tagRepairedCards),
            state = state,
            activeCardsLabel = SettingsTextCopy.activeCardsLabel(),
            suspendedCardsLabel = SettingsTextCopy.suspendedCardsLabel(),
            taggedCardsLabel = SettingsTextCopy.taggedCardsLabel(),
            weakCardsLabel = SettingsTextCopy.weakCardsLabel(),
            browserQueryCardsLabel = SettingsTextCopy.browserQueryLabel(),
            browserQueryLabel = SettingsTextCopy.ankiBrowserQueryLabel(),
            browserQueryHint = SettingsTextCopy.ankiBrowserQueryHint(),
            browserQueryHelperText = SettingsTextCopy.ankiBrowserQueryHelperText(),
            tagsLabel = SettingsTextCopy.ankiNoteTagsLabel(),
            tagsHint = SettingsTextCopy.ankiNoteTagsHint(),
            difficultyLabel = SettingsTextCopy.fsrsDifficultyLabel(),
            lapsesLabel = SettingsTextCopy.lapsesLabel(),
            minMatchingLabel = SettingsTextCopy.minimumMatchingCardsLabel(),
            saveLabel = SettingsTextCopy.saveImportFiltersLabel(),
            onSave = SettingsImportFilterSaveAction(::saveImportFilters),
            tagRepairedCardsLabel = SettingsTextCopy.tagRepairedCardsLabel(),
        )
    }

    private fun presetButtons(tagRepairedCards: Boolean): List<SettingsImportPresetButtonModel> {
        return SettingsImportPreset.defaults().map { preset ->
            SettingsImportPresetButtonModel(
                label = preset.label(),
                onClick = SettingsImportFilterAction {
                    applyPreset(preset, tagRepairedCards)
                }
            )
        }
    }

    private fun saveImportFilters(state: SettingsImportFiltersState) {
        val parsedTags = RecordsBase.parseImportTags(state.tags)
        val queryText = state.browserQuery.trim()
        if (state.browserQueryCards && queryText.isEmpty()) {
            Toast.makeText(activity, SettingsTextCopy.browserQueryRequiredToast(), Toast.LENGTH_SHORT).show()
            return
        }
        if (!validation.hasSelectedImportSource(
                state.activeCards,
                state.suspendedCards,
                state.taggedCards,
                state.weakCards,
                state.browserQueryCards,
                parsedTags,
                queryText
            )
        ) {
            Toast.makeText(activity, SettingsTextCopy.importSourceRequiredToast(), Toast.LENGTH_SHORT).show()
            return
        }
        val parsedThresholds = validation.readImportThresholds(
            state.difficulty,
            state.lapses,
            state.minMatching
        ) ?: return
        activity.runSettingsWrite(
            traceSection = "kani.settings.import-filters.save",
            write = {
                activity.saveSettings(
                    SettingsSaveCommand.ImportFilters(
                        activeCards = state.activeCards,
                        suspendedCards = state.suspendedCards,
                        taggedCards = state.taggedCards,
                        tags = parsedTags.joinToString(" "),
                        weakCards = state.weakCards,
                        weakDifficulty = parsedThresholds.difficulty,
                        weakLapses = parsedThresholds.lapseThreshold,
                        minMatchingCards = parsedThresholds.minCards,
                        browserQueryCards = state.browserQueryCards,
                        browserQuery = queryText,
                        tagRepairedCards = state.tagRepairedCards,
                    ),
                )
            },
        ) {
            Toast.makeText(activity, SettingsTextCopy.importFiltersSavedToast(), Toast.LENGTH_LONG).show()
            activity.renderSettingsImportSync(true)
        }
    }

    private fun applyPreset(
        preset: SettingsImportPreset,
        tagRepairedCards: Boolean,
    ) {
        activity.runSettingsWrite(
            traceSection = "kani.settings.import-filters.preset",
            write = {
                activity.saveSettings(
                    SettingsSaveCommand.ImportFilters(
                        activeCards = preset.activeCards(),
                        suspendedCards = preset.suspendedCards(),
                        taggedCards = preset.taggedCards(),
                        tags = preset.tags(),
                        weakCards = preset.weakCards(),
                        weakDifficulty = preset.weakDifficulty(),
                        weakLapses = preset.weakLapses(),
                        minMatchingCards = preset.minMatchingCards(),
                        browserQueryCards = preset.browserQueryCards(),
                        browserQuery = preset.browserQuery(),
                        tagRepairedCards = tagRepairedCards,
                    ),
                )
            },
        ) {
            Toast.makeText(activity, SettingsTextCopy.importPresetSavedToast(), Toast.LENGTH_LONG).show()
            activity.renderSettingsImportSync(true)
        }
    }

    private companion object {
        fun trimmed(value: String?): String = value?.trim().orEmpty()

        fun decimalText(value: Double): String = String.format(Locale.ROOT, "%.1f", value)

        fun thresholdText(value: Int): String = value.coerceAtLeast(1).toString()
    }
}
